api/nodes/views.py

```diff
@@ -106,8 +145,125 @@ def get_queryset(self):
+        ContributorOrPublic,
+    )
+
+    def get_valid_self_link_methods(self, root_folder=False):
+        valid_methods = {'file': ['GET'], 'folder': [], }
+        user = self.request.user
+        if user is None or user.is_anonymous():
+            return valid_methods
+
+        permissions = self.get_node().get_permissions(user)
+        if 'write' in permissions:
+            valid_methods['file'].append('POST')
+            valid_methods['file'].append('DELETE')
+            valid_methods['folder'].append('POST')
+            if not root_folder:
+                valid_methods['folder'].append('DELETE')
+
+        return valid_methods
+
+    def get_file_item(self, item, cookie, obj_args):
+        file_item = {
+            'valid_self_link_methods': self.get_valid_self_link_methods()[item['kind']],
+            'provider': item['provider'],
+            'path': item['path'],
+            'name': item['name'],
+            'node_id': self.get_node()._primary_key,
+            'cookie': cookie,
+            'args': obj_args,
+            'waterbutler_type': 'file',
+            'item_type': item['kind'],
+        }
+        if file_item['item_type'] == 'folder':
+            file_item['metadata'] = {}
+        else:
+            file_item['metadata'] = {
+                'content_type': item['contentType'],
+                'modified': item['modified'],
+                'size': item['size'],
+                'extra': item['extra'],
+            }
+        return file_item
+
+    def get_queryset(self):
+        query_params = self.request.query_params
+
+        addons = self.get_node().get_addons()
+        user = self.request.user
+        cookie = None if self.request.user.is_anonymous() else user.get_or_create_cookie()
+        node_id = self.get_node()._id
+        obj_args = self.request.parser_context['args']
+
+        provider = query_params['provider'] if 'provider' in query_params else None
+        path = query_params['path'] if 'path' in query_params else '/'
+        files = []
+
+        if provider is None:
+            valid_self_link_methods = self.get_valid_self_link_methods(True)
+            for addon in addons:
+                if addon.config.has_hgrid_files:
+                    files.append({
+                        'valid_self_link_methods': valid_self_link_methods['folder'],
+                        'provider': addon.config.short_name,
+                        'name': addon.config.short_name,
+                        'path': path,
+                        'node_id': node_id,
+                        'cookie': cookie,
+                        'args': obj_args,
+                        'waterbutler_type': 'data',
+                        'item_type': 'folder',
+                        'metadata': {},
+                    })
+        else:
+            url = waterbutler_url_for('data', provider, path, self.kwargs['pk'], node_id, obj_args)
+            waterbutler_request = requests.get(url)
```

Comment:
This needs error checking and reporting to the user in the case of 400 series errors.